### PR TITLE
Users' quote verification

### PR DESCRIPTION
**Pull request by jumaffre**
Resolves https://github.com/microsoft/CCF/issues/926

Building on https://github.com/microsoft/CCF/pull/962, this PR allows application endpoints to verify a quote supplied by an unknown (i.e. not registered in the `ccf.users` table) client against 1) the client certificate supplied by the TLS connection and 2) a list of trusted enclave measurements, stored in the new `ccf.users.code_ids` table. 

Assuming that clients are running in enclaves, this lets users self-register without going through a full round of governance to add a new user, as long as new users can provide evidence that they are running in a trusted enclave.

Since the solution is very much like the join protocol, I have refactored `quoteverification.h` (now `quote.h` as I moved quote generation code from `nodestate.h` in there too).

The aim is not to test this end-to-end (this would require setting up client enclaves, etc.) but provide the building blocks for app developers to do that.

### CODE REVIEW

File: src/apps/logging/logging.cpp [hunk 1]
```diff
@@ -196,21 +199,143 @@ namespace ccfapp
...
-          auto view = tx.get_view(records);
-          view->put(in.id, log_line);
-          return make_success(true);
-        };
+      auto log_record_anonymous = [this](
+                                    RequestArgs& args,
+                                    nlohmann::json&& params) {
+        const auto in = params.get<LoggingRecord::In>();
+
+        std::vector<uint8_t> raw_quote = tls::raw_from_b64(
```
Comment (jumaffre) [inline src/apps/logging/logging.cpp]: Obviously, all of this code will go before I merge this. This is a quote I self-generated on another node and extracted the code id from, just to prove myself that it works.

Comment (jumaffre) [reply src/apps/logging/logging.cpp]: * In practice, the quote would be passed as as argument to the RPC.

**Comment by cimetrics** [comment]
self_register_users@[6168 aka 20200320.18](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=6168&view=results) vs master ewma over 30 builds from [5890](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=5890&view=results) to [6160](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=6160&view=results)
images

File: src/apps/logging/logging.cpp [hunk 2]
```diff
@@ -196,21 +199,144 @@ namespace ccfapp
...
+          tls::cert_der_to_pem(args.rpc_ctx->session->caller_cert));
+        if (rc != QuoteVerificationResult::VERIFIED)
+        {
+          const auto [code, message] =
+            QuoteVerifier::quote_verification_error(rc);
+          return make_error(code, message);
+        }
+#endif
+
+        if (in.msg.empty())
```
Comment (olgavrou) [inline src/apps/logging/logging.cpp]: could be nice to first check if the msg can be recorded and then check the quote?

Comment (jumaffre) [reply src/apps/logging/logging.cpp]: It's generally good practice to first check authZ and then enforce application-specific checks. This code will go before I merge this PR anyway.

**Pull request merged**